crates/binjs_io/src/context/strings.rs

```diff
@@ -0,0 +1,178 @@
+    pub fn strings(&self) -> &[String] {
+        self.strings.as_ref()
+    }
+
+    /// Read a string table from a stream.
+    ///
+    /// Consumes exactly the bytes used to define the string table.
+    pub fn read<R: Read>(input: &mut R) -> Result<Self, Error> {
+        let string_count = input.read_varu32_no_normalization()?.value;
+
+        // The strings we decode, in the order in which we decode them.
+        let mut strings = Vec::new();
+
+        for _ in 0..string_count {
+            let mut chars = Vec::new();
+            // Read bytes one by one to handle escapes.
+            loop {
+                let mut buf = [0];
+                input.read_exact(&mut buf)?;
+
+                let byte = match buf[0] {
+                    0 => {
+                        // NUL terminator.
+                        break;
+                    }
+                    1 => {
+                        // Escaped char.
+                        input.read_exact(&mut buf)?;
+                        match buf[0] {
+                            0 => 0,
+                            1 => 1,
+                            _ => return Err(Error::new(ErrorKind::InvalidData, "Invalid escape")),
+                        }
+                    }
+                    byte => byte,
+                };
+                chars.push(byte);
+            }
+            // Attempt to decode string as UTF-8.
+            let string = String::from_utf8(chars)
+                .or_else(|_| Err(Error::new(ErrorKind::InvalidData, "Invalid utf-8")))?;
+            strings.push(string);
+        }
+
+        Ok(StringTable { strings })
+    }
+}
+
+pub struct ExternalStringDictionary {
+    strings: StringTable,
+}
+
+impl ExternalStringDictionary {
+    /// The strings in this dictionary, in the order in which they appear in the stream.
+    pub fn strings(&self) -> &[String] {
+        self.strings.strings()
+    }
+
+    /// Read a string dictionary from a stream.
+    ///
+    /// Consumes exactly the bytes used to define the string dictionary.
+    pub fn read<R: Read>(input: &mut R) -> Result<Self, Error> {
+        // Check magic header
+        let mut header_bytes = [0; 7];
+        assert_eq!(DICTIONARY_MAGIC_HEADER.len(), header_bytes.len());
+
+        input.read_exact(&mut header_bytes)?;
+        if &header_bytes != DICTIONARY_MAGIC_HEADER {
+            return Err(Error::new(
+                ErrorKind::InvalidData,
+                "Format error: Not an external strings dictionary",
+            ));
+        }
+
+        // Read contents
+        let strings = StringTable::read(input)?;
+        Ok(ExternalStringDictionary { strings })
+    }
+}
+
+#[test]
+fn test_empty_external_string_dictionary() {
+    let bytes = b"astdict\0";
+    let mut input = std::io::Cursor::new(&bytes);
+    let dictionary = ExternalStringDictionary::read(&mut input).unwrap();
+    assert_eq!(dictionary.strings().len(), 0);
+}
+
+// ----- Writing
+
+impl StringTable {
```

Comment:
In practice you might decode many files with one external dictionary. Does it make sense to have a string table point to a couple of tables and handle concatenating them virtually?

Reply:
Yes, that's what I intend to do, but not in this module.